server/src/main/java/io/crate/planner/optimizer/matcher/WithPattern.java

```diff
@@ -37,12 +40,13 @@
     }
 
     @Override
-    public Match<T> accept(Object object, Captures captures) {
-        Match<T> match = firstPattern.accept(object, captures);
+    public Match<T> accept(Object object, Captures captures, Function<LogicalPlan, LogicalPlan> resolvePlan) {
+        Match<T> match = firstPattern.accept(object, captures, resolvePlan);
         return match.flatMap(matchedValue -> {
-            Optional<?> optProperty = getProperty.apply(matchedValue);
+            Optional<?> optProperty = getProperty.apply(matchedValue)
+                .map(value -> value instanceof GroupReference groupReference ? resolvePlan.apply(groupReference) : value);
```

Comment:
Couldn`t the `resolvePlan` implementation do the instanceof check to decouple `WithPattern` from the `GroupReference` ?

Otherwise it leaks part of the iterative/memo implementation into the pattern.

Reply:
Then I would need to cast Object to LogicalPlan to make acceptable in resolvePlan. Do you prefer this ?

Reply:
I could match against LogicalPlan instead of GroupReference or we could make resolvePlan `Function<Object, LogicalPlan>`.